### DISCUSSION TITLE
AR Button not responding embed code in Native Apps webview on IOS devices.

### DISCUSSION THREAD

**Discussion by base-krchome** [google/model-viewer#2615] [Q&A]
### Description

AR Button not responding(not work) embed code in Native Apps for Ios devices.

AR button works in embed code which includes model viewer in  Native App webview for Android Devices, but not for IOS.

Topics opened for that but it closed without solution.

My opinion this issue about native app codes.

My observation :  

Model viewer ar button **works** on ios devices in Twitter app
Model viewer ar button **not works** on ios devices in Instagram app
Model viewer ar button **not works** on ios devices in Linkedln app

So it should be Native app code issue.

Can I fix this issue with javascript code in modelviewer?
Should we add quicklook library to Native app?

Can you show the true way for me.

Thanks for help

#### Live Demo
All model viewer demos.

### Version
All
- model-viewer: vX.X.X

### Browser Affected

- [ ] Chrome, version: xx.x.xxxx.xx
- [ ] Edge
- [ ] Firefox
- [ ] IE
- [ ] Safari

### OS

- [ ] Android
- [ + ] iOS
- [ ] Linux
- [ ] MacOS
- [ ] Windows

### AR

- [ ] WebXR
- [ ] SceneViewer
- [+ ] QuickLook

**Comment by elalish** [comment]
I'm not sure it's possible to launch Quick Look from the iOS Instagram or Twitter apps. Has anyone attempted this directly? If it is possible and we're simply screening for the wrong user agent string or something, that would be a simple fix. However, I have no experience with those apps, so any information you can give would be a great help.

**Comment by base-krchome** [comment]
we send message "modelviewer AR link" to each other on these apps. Then we opened. These apps open the link without leaving the app(Webview)

On Android devices, everything works fine.
ON IOS devices,

AR button works on Twitter webview..
but not on instagram. 

So twitter someway opens the quicklook in their webview but instagram doesn't..

**Comment by canberka** [comment]
<h4 style="color: rgb(57, 61, 61); font-family: &quot;Open Sans&quot;, sans-serif; font-size: 18px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Facebook for iOS User-Agent string</h4><div class="ua-string" style="color: rgb(97, 100, 100); font-family: &quot;Open Sans&quot;, sans-serif; font-size: 18px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div id="crayon-60ba78d019b51915758855" class="crayon-syntax crayon-theme-sublime-text-copy crayon-font-monaco crayon-os-pc print-yes notranslate crayon-wrapped" data-settings=" minimize scroll-mouseover disable-anim wrap" style="margin: 15px 0px; border-width: 1px !important; border-color: rgb(153, 153, 153) !important; border-style: solid !important; text-shadow: none !important; background: rgb(39, 40, 34) !important; text-align: left; box-sizing: border-box; text-size-adjust: none; overflow: hidden !important; position: relative !important; direction: ltr !important; width: 766.156px; padding: 0px; font-family: Monaco, MonacoRegular, &quot;Courier New&quot;, monospace; font-weight: 500; height: auto;"><div class="crayon-plain-wrap" style="line-height: 1.3em; font-family: Monaco, MonacoRegular, &quot;Courier New&quot;, monospace !important; background: 0px center; border: 0px; padding: 0px !important; margin: 0px !important; text-align: left; height: auto !important;"></div><div class="crayon-main" style="line-height: 1.3em; font-family: Monaco, MonacoRegular, &quot;Courier New&quot;, monospace !important; background: 0px center; border: 0px; padding: 0px; margin: 0px; text-align: left; width: 764.156px; overflow: auto; position: relative; z-index: 1;">
1 | Mozilla/5.0 (iPhone; CPU iPhone OS 8_2 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) Mobile/12D508 [FBAN/FBIOS;FBAV/27.0.0.10.12;FBBV/8291884;FBDV/iPhone7,1;FBMD/iPhone;FBSN/iPhone OS;FBSV/8.2;FBSS/3; FBCR/vodafoneIE;FBID/phone;FBLC/en_US;FBOP/5]
-- | --


</div></div></div>




<h4 style="color: rgb(57, 61, 61); font-family: &quot;Open Sans&quot;, sans-serif; font-size: 18px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Twitter for iOS User-Agent string</h4><div class="ua-string" style="color: rgb(97, 100, 100); font-family: &quot;Open Sans&quot;, sans-serif; font-size: 18px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div id="crayon-60ba78d019b52693664870" class="crayon-syntax crayon-theme-sublime-text-copy crayon-font-monaco crayon-os-pc print-yes notranslate crayon-wrapped" data-settings=" minimize scroll-mouseover disable-anim wrap" style="margin: 15px 0px; border-width: 1px !important; border-color: rgb(153, 153, 153) !important; border-style: solid !important; text-shadow: none !important; background: rgb(39, 40, 34) !important; text-align: left; box-sizing: border-box; text-size-adjust: none; overflow: hidden !important; position: relative !important; direction: ltr !important; width: 766.156px; padding: 0px; font-family: Monaco, MonacoRegular, &quot;Courier New&quot;, monospace; font-weight: 500; height: auto;"><div class="crayon-plain-wrap" style="line-height: 1.3em; font-family: Monaco, MonacoRegular, &quot;Courier New&quot;, monospace !important; background: 0px center; border: 0px; padding: 0px !important; margin: 0px !important; text-align: left; height: auto !important;"></div><div class="crayon-main" style="line-height: 1.3em; font-family: Monaco, MonacoRegular, &quot;Courier New&quot;, monospace !important; background: 0px center; border: 0px; padding: 0px; margin: 0px; text-align: left; width: 764.156px; overflow: hidden; position: relative; z-index: 1;">
1 | Mozilla/5.0 (iPhone; CPU iPhone OS 8_2 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) Mobile/12D508
-- | --


</div></div></div>



<h4 style="color: rgb(57, 61, 61); font-family: &quot;Open Sans&quot;, sans-serif; font-size: 18px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Safari for iOS User-Agent string</h4><div class="ua-string" style="color: rgb(97, 100, 100); font-family: &quot;Open Sans&quot;, sans-serif; font-size: 18px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div id="crayon-60ba78d019b55010530266" class="crayon-syntax crayon-theme-sublime-text-copy crayon-font-monaco crayon-os-pc print-yes notranslate crayon-wrapped" data-settings=" minimize scroll-mouseover disable-anim wrap" style="margin: 15px 0px; border-width: 1px !important; border-color: rgb(153, 153, 153) !important; border-style: solid !important; text-shadow: none !important; background: rgb(39, 40, 34) !important; text-align: left; box-sizing: border-box; text-size-adjust: none; overflow: hidden !important; position: relative !important; direction: ltr !important; width: 766.156px; padding: 0px; font-family: Monaco, MonacoRegular, &quot;Courier New&quot;, monospace; font-weight: 500; height: auto;"><div class="crayon-plain-wrap" style="line-height: 1.3em; font-family: Monaco, MonacoRegular, &quot;Courier New&quot;, monospace !important; background: 0px center; border: 0px; padding: 0px !important; margin: 0px !important; text-align: left; height: auto !important;"></div><div class="crayon-main" style="line-height: 1.3em; font-family: Monaco, MonacoRegular, &quot;Courier New&quot;, monospace !important; background: 0px center; border: 0px; padding: 0px; margin: 0px; text-align: left; width: 764.156px; overflow: auto; position: relative; z-index: 1;">
1 | Mozilla/5.0 (iPhone; CPU iPhone OS 8_2 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) Version/8.0 Mobile/12D508
-- | --


</div></div></div>



So twitter's user agent looks like Safari, thats why it works... I dont know @elalish  and model viewer team can handle with it

**Comment by elalish** [comment]
Hmm, I take it back; we're not narrowing it down by user agent string in this case. I believe [this](https://github.com/google/model-viewer/blob/master/packages/model-viewer/src/constants.ts#L71) is what's failing, in which case the browser really doesn't support AR, though I don't know why. Old version? If anyone can debug and verify that, it would be great. Or really any kind of repro steps to test the facebook browser.

**Comment by sercandemirtas** [comment]
Here you go.. @elalish @canberka 

"EVİNDE GÖRÜNTÜLE" means view in your space..(AR button)

Button disappears and comes back when clicked.

This is debug; ("Touch down at" is when we clicked button)

Path 1:
	Touch down at 222.0, 589.0, offset=0.00s
	Touch up at 222.0, 589.0, offset=0.10s.
Jun  7 21:19:01 IPhone-11-Pro-IO-UK-10 backboardd(CoreBrightness)[66] <Error>: invalid parameters
Jun  7 21:19:01 IPhone-11-Pro-IO-UK-10 Karaca(WebKit)[38474] <Notice>: Single tap identified. Request details on potential zoom. (0x11b89fa00)
Jun  7 21:19:01 IPhone-11-Pro-IO-UK-10 nearbyd[19073] <Error>: onHIDEvent HIDServiceClient is nil
<XCSynthesizedEventRecord 'tap'>
Path 1:


https://user-images.githubusercontent.com/68244546/121079524-ea28a280-c7e2-11eb-83cb-27e204e0b997.mp4

and this is app webview for ios...

![ourapp-agent](https://user-images.githubusercontent.com/68244546/121079611-04fb1700-c7e3-11eb-9ea6-64b46eda0e72.png)


I hope this helps

**Comment by canberka** [comment]
Hmm, maybe it is all about privacy  for not open the AR, as same as webxr not opened in wkwebview.

**Comment by sprengerst** [comment]
👍🏻+1 for a solution. :)

**Comment by canberka** [comment]
there is no solution for Wkwebview, only works UIWebview with SafariServices @sprengerst which is works on Twitter

**Comment by sprengerst** [comment]
@canberka Thanks for theinfo. Is there any event that is triggered the browser is not capable of launching AR? The "failed" event doesn't seem to be triggered.
